### PR TITLE
Import AddressBase to offline table and then swap the table names

### PR DESCRIPTION
**Pull request by aldavidson**
AddressBaseImporter now:
- imports all given csv files in one go into an offline table with no indexes
- copies index defs over from the live table
- swaps the live & offline table names and renames the indexes to be consistent

This last stop is done in one transaction (Postgres has transactional schema changes), so the live table is completely unaffected by an ongoing import, and the name swap happens atomically.

The full import of ~29m rows now runs in ~39mins on my macbookpro, although it will be slower on live

### CODE REVIEW

File: scripts/addressbase_import.sh [hunk 1]
```diff
@@ -17,13 +26,32 @@ function import_file {
...
+#   table name to replace in index names
+#   table name to substitute into index names
+function rename_indexes_sql {
+  exec_sql "select indexdef from pg_indexes where tablename = '$1'" \
+  | sed 's/.*INDEX \([^ ]*\).*/ALTER INDEX \1 RENAME TO ZZZ_\1;/' \
+  | sed "s/ZZZ_${2}_/${3}_/"
+}
+
 # read -d'' reads multiple lines from stdin ignoring newlines
 # see http://serverfault.com/a/72511 for more details of how this works
```
Comment (ashb) [inline scripts/addressbase_import.sh]: This comment doesn't make sense in since the `read` has been removed

File: scripts/addressbase_import.sh [hunk 2]
```diff
@@ -51,22 +79,29 @@ read -d '' CREATE_TABLE_SQL <<"EOF"
     PROCESS_DATE date
   );
 
-  TRUNCATE TABLE tmp_addressbase_import;
```
Comment (ashb) [inline scripts/addressbase_import.sh]: We don't need a truncate anymore as we start with DROP TABLE IF EXISTS

File: scripts/addressbase_import.sh [hunk 3]
```diff
@@ -120,28 +155,54 @@ read -d '' CONVERT_DATA_SQL <<"EOF"
...
+    import_file $filename
+  done
 
-echo "creating temporary import table"
-exec_sql "$CREATE_TABLE_SQL"
+  exec_sql "SELECT CHANGE_TYPE, COUNT(*) AS num_to_import FROM $TEMP_TABLE_NAME GROUP BY CHANGE_TYPE;"
+  echo "converting data"
+  # NOTE: the rename_indexes_sql must be generated AFTER
+  # the cleanup_tables_sql has been run!
+  exec_in_transaction "$CONVERT_DATA_SQL" && \
```
Comment (ashb) [inline scripts/addressbase_import.sh]: for consistency could you make this a function too?

File: scripts/addressbase_import.sh (hunk 1 above)
Comment (aldavidson) [inline scripts/addressbase_import.sh]: good spot - it's gone

File: scripts/addressbase_import.sh (hunk 2 above)
Comment (aldavidson) [inline scripts/addressbase_import.sh]: it helps keep storage usage down if we don't keep this temporary copy around for any more time than it's needed. It's purely a database-server-side mirror of the csv files, so not primary data, and a TRUNCATE takes less than a second.

**Comment by aldavidson** [comment]
squished the commit history down to two logical steps.

**Comment by ashb** [comment]
Thanks Al! :sparkle:

**Pull request merged**